src/csharp/Grpc.Core/GrpcEnvironment.cs

```diff
@@ -360,12 +361,26 @@ internal static string GetCoreVersionString()
 
         internal static void GrpcNativeInit()
         {
+            if (!IsNativeShutdownAllowed && alreadyInvokedNativeInit)
+            {
+                // Normally grpc_init and grpc_shutdown calls should come in pairs (C core does reference counting),
+                // but in case we avoid grpc_shutdown calls altogether, calling grpc_init has no effect
+                // besides incrementing an internal C core counter that could theoretically overflow.
+                // NOTE: synchronization not necessary here as we are only trying to avoid calling grpc_init
```

Comment:
is the idea here that a race won't happen because it will never overflow quickly enough? In that case, even so, I think it would make it easier to read if `alreadyInvokedNativeInit` was under a lock

Reply:
The idea is that it's not guaranteed that if we set alreadyInvokedNativeInit, other thread will immediately see that, but they will see it "very soon" and we don't mind invoking grpc_init once or twice in the meantime. In the worst case, each thread will run grpc_init once and that's still far from risking the owerflow.
The idea is the same as Joshua Bloch's  "Racy-Single-Check" idiom - 
http://javaagile.blogspot.com/2013/05/the-racy-single-check-idiom.html
which is used e.g. in Java String to initialize the hashCode field.

But you're right that this technique is a bit of an overkill for this situation. I'll double check that GrpcNativeInit only happens when a GrpcEnvironment gets initialized and probably use a lock.

Reply:
I'd tend to err on the side of caution around this. And I really have no idea if this is an issue in C#, but see [benign data races blog post](https://software.intel.com/en-us/blogs/2013/01/06/benign-data-races-what-could-possibly-go-wrong)